utils/s2n_random.c

```diff
@@ -128,65 +130,73 @@ S2N_RESULT s2n_get_mix_entropy(struct s2n_blob *blob)
     return S2N_RESULT_OK;
 }
 
-void s2n_on_fork(void)
-{
-    zero_if_forked = 0;
-}
-
-static inline S2N_RESULT s2n_defend_if_forked(void)
+static S2N_RESULT s2n_initialise_drbgs(void)
 {
     uint8_t s2n_public_drbg[] = "s2n public drbg";
     uint8_t s2n_private_drbg[] = "s2n private drbg";
-    struct s2n_blob public = {.data = s2n_public_drbg,.size = sizeof(s2n_public_drbg) };
-    struct s2n_blob private = {.data = s2n_private_drbg,.size = sizeof(s2n_private_drbg) };
+    struct s2n_blob public = { .data = s2n_public_drbg, .size = sizeof(s2n_public_drbg) };
+    struct s2n_blob private = { .data = s2n_private_drbg, .size = sizeof(s2n_private_drbg) };
 
-    if (zero_if_forked == 0) {
-        /* Clean up the old drbg first */
-        RESULT_GUARD(s2n_rand_cleanup_thread());
-        /* Instantiate the new ones */
-        RESULT_GUARD(s2n_drbg_instantiate(&per_thread_public_drbg, &public, S2N_AES_128_CTR_NO_DF_PR));
-        RESULT_GUARD(s2n_drbg_instantiate(&per_thread_private_drbg, &private, S2N_AES_128_CTR_NO_DF_PR));
-        zero_if_forked_ptr = zeroed_when_forked_page;
-        zero_if_forked = 1;
-    }
+    RESULT_GUARD(s2n_drbg_instantiate(&s2n_per_thread_rand_state.public_drbg, &public, S2N_AES_128_CTR_NO_DF_PR));
+    RESULT_GUARD(s2n_drbg_instantiate(&s2n_per_thread_rand_state.private_drbg, &private, S2N_AES_256_CTR_NO_DF_PR));
+
+    s2n_per_thread_rand_state.drbgs_initialised = true;
 
     return S2N_RESULT_OK;
 }
 
-S2N_RESULT s2n_get_public_random_data(struct s2n_blob *blob)
+static S2N_RESULT s2n_maybe_initialise_drbgs(void)
 {
-    RESULT_GUARD(s2n_defend_if_forked());
+    if (s2n_per_thread_rand_state.drbgs_initialised == false) {
+        RESULT_GUARD(s2n_initialise_drbgs());
 
-    uint32_t offset = 0;
-    uint32_t remaining = blob->size;
+        /* Then cache the fork generation number. We just initialised the drbg
+         * states with new entropy and forking is not an external event.
+         */
+        uint64_t returned_fork_generation_number = 0;
+        RESULT_GUARD(s2n_get_fork_generation_number(&returned_fork_generation_number));
+        s2n_per_thread_rand_state.cached_fork_generation_number = returned_fork_generation_number;
+    }
 
-    while(remaining) {
-        struct s2n_blob slice = { 0 };
+    return S2N_RESULT_OK;
+}
 
-        RESULT_GUARD_POSIX(s2n_blob_slice(blob, &slice, offset, MIN(remaining, S2N_DRBG_GENERATE_LIMIT)));;
+/* s2n_defend_against_ube() implements defenses against ube's (uniqueness
+ * breaking events). Currently, only implements fork detection.
+ */
+static S2N_RESULT s2n_defend_against_ube(void)
```

Comment:
I'm sorry, but "ube" is just a funny name for a serious thing xD I couldn't find any relevant google results for "ube drgb", so I'm not sure that this is a common acronym. Maybe "s2n_defend_uniqueness" or "s2n_ensure_uniqueness"?

Reply:
Sorry, couldn't help dropping this link: https://www.eatingwell.com/article/7835700/what-is-ube/